### PR TITLE
Feature: crm_resource: Add the --element option for --get-parameter

### PR DESCRIPTION
**Pull request by waltdisgrace**
Here is an example of --get-parameter being used with --element:

```
[root@pcmk-1 ~]# crm_resource -r myping --get-parameter description --element
test_description
```

### CODE REVIEW

File: tools/crm_resource.c [hunk 1]
```diff
@@ -592,6 +592,10 @@ static GOptionEntry addl_entries[] = {
       "Use resource utilization attribute instead of instance attribute\n"
       INDENT "(with -p, -g, -d)",
       NULL },
+    { "element", 'y', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, attr_set_type_cb,
```
Comment (kgaillot) [inline tools/crm_resource.c:595]: I'm thinking we don't need a short option for this

File: tools/crm_resource.c [hunk 2]
```diff
@@ -671,7 +675,8 @@ attr_set_type_cb(const gchar *option_name, const gchar *optarg, gpointer data, G
         options.attr_set_type = XML_TAG_META_SETS;
     } else if (pcmk__str_any_of(option_name, "-z", "--utilization", NULL)) {
         options.attr_set_type = XML_TAG_UTILIZATION;
-    }
+    } else if (pcmk__str_any_of(option_name, "-y", "--element", NULL)) {
+        options.attr_set_type = XML_TAG_ELEMENT;
```
Comment (kgaillot) [inline tools/crm_resource.c:679]: This won't be an actual XML tag, just a magic string that doesn't match any real tag, so you can just define a constant for it in this file

File: tools/crm_resource.c [hunk 3]
```diff
@@ -2031,6 +2036,15 @@ main(int argc, char **argv)
...
 
+            } else if (pcmk__str_eq(options.attr_set_type, XML_TAG_ELEMENT, pcmk__str_casei)) {
+                params = pcmk__strkey_table(free, free);
+
+                // TODO:
+                // - Get element attributes into a hash table, similar to get_meta_attributes above.
+                // - The table maps attribute names to values.
+                // - For example, description maps to "test_description"
+                // - Then I am done.
+
```
Comment (kgaillot) [inline tools/crm_resource.c:2047]: Rather than create a hash table, it might be easier just to iterate through the attributes on rsc->xml (see existing loops using pcmk__xe_first_attr() and xmlAttr)

**Review by kgaillot — COMMENTED** [review]
good start

**Review by kgaillot — COMMENTED** [review]
Looking good. Also add cts-cli regression tests for existing and nonexisting attributes

File: tools/crm_resource.c [hunk 4]
```diff
@@ -2031,15 +2037,34 @@ main(int argc, char **argv)
...
             } else {
                 params = pcmk__strkey_table(free, free);
                 pe__unpack_dataset_nvpairs(rsc->xml, XML_TAG_UTILIZATION, NULL, params,
                                            NULL, FALSE, data_set);
             }
 
-            rc = out->message(out, "attribute-list", rsc, options.prop_name, params);
-            if (free_params) {
-                g_hash_table_destroy(params);
+            if (params) {
```
Comment (nrwahl2) [inline tools/crm_resource.c:2063]: For clarity about each variable's data type, we avoid implicit booleans in conditions in new code. We like to explicitly compare against NULL (`if (params != NULL)`), or against 0 if the variable is an integer.

That guideline is here: https://clusterlabs.org/pacemaker/doc/2.1/Pacemaker_Development/singlehtml/#control-statements-if-else-while-for-switch

File: tools/crm_resource_print.c [hunk 5]
```diff
@@ -131,27 +131,22 @@ cli_resource_print(pe_resource_t *rsc, pe_working_set_t *data_set, bool expanded
...
 }
 
-PCMK__OUTPUT_ARGS("attribute-list", "pe_resource_t *", "char *", "GHashTable *")
+PCMK__OUTPUT_ARGS("attribute-list", "pe_resource_t *", "char *", "char *")
 static int
 attribute_list_default(pcmk__output_t *out, va_list args) {
     pe_resource_t *rsc = va_arg(args, pe_resource_t *);
     char *attr = va_arg(args, char *);
-    GHashTable *params = va_arg(args, GHashTable *);
+    char *value = va_arg(args, char *);
```
Comment (kgaillot) [inline tools/crm_resource_print.c:139]: This and the above should be const char * (same for the text implementation)

File: tools/crm_resource_print.c [hunk 6]
```diff
@@ -131,27 +131,22 @@ cli_resource_print(pe_resource_t *rsc, pe_working_set_t *data_set, bool expanded
...
-        value = g_hash_table_lookup(params, attr);
-    }
     if (value != NULL) {
         out->begin_list(out, NULL, NULL, "Attributes");
         out->list_item(out, attr, "%s", value);
         out->end_list(out);
+        return pcmk_rc_ok;
     } else {
         out->err(out, "Attribute '%s' not found for '%s'", attr, rsc->id);
+        return pcmk_rc_unknown_format;
```
Comment (kgaillot) [inline tools/crm_resource_print.c:148]: It does make sense to return a different code for unknown attributes, but I wouldn't want to change the behavior now for backward compatibility reasons (same for text implementation). I think the idea behind the original code was that even if there is no such attribute, we successfully determined that.

**Review by kgaillot — COMMENTED** [review]
looks great :)

File: tools/crm_resource_print.c [hunk 7]
```diff
@@ -131,26 +131,21 @@ cli_resource_print(pe_resource_t *rsc, pe_working_set_t *data_set, bool expanded
     return pcmk_rc_ok;
 }
 
-PCMK__OUTPUT_ARGS("attribute-list", "pe_resource_t *", "char *", "GHashTable *")
+PCMK__OUTPUT_ARGS("attribute-list", "pe_resource_t *", "char *", "char *")
```
Comment (kgaillot) [inline tools/crm_resource_print.c:134]: This should be `const char *`. Basically it indicates that this function won't change whatever value is passed in.

(The first `char *` could be const as well, but we don't have to change it now. If you want to, do it in another commit, and update the `va_arg()` line as well, and do the same in `attribute_list_text()`.)

File: tools/crm_resource_print.c [hunk 8]
```diff
@@ -131,26 +131,21 @@ cli_resource_print(pe_resource_t *rsc, pe_working_set_t *data_set, bool expanded
...
 static int
 attribute_list_default(pcmk__output_t *out, va_list args) {
     pe_resource_t *rsc = va_arg(args, pe_resource_t *);
     char *attr = va_arg(args, char *);
-    GHashTable *params = va_arg(args, GHashTable *);
+    const char *value = va_arg(args, char *);
```
Comment (kgaillot) [inline tools/crm_resource_print.c:139]: The second char here should be `const char *` too

File: tools/crm_resource_print.c [hunk 9]
```diff
@@ -229,27 +224,21 @@ agent_status_xml(pcmk__output_t *out, va_list args) {
     return pcmk_rc_ok;
 }
 
-PCMK__OUTPUT_ARGS("attribute-list", "pe_resource_t *", "char *", "GHashTable *")
+PCMK__OUTPUT_ARGS("attribute-list", "pe_resource_t *", "char *", "char *")
```
Comment (kgaillot) [inline tools/crm_resource_print.c:227]: This and the `va_arg()` line need to be const as well

**Review by kgaillot — COMMENTED** [review]
Looks great, just about done
